// gameObjects/effects/EnergyBubblePop.js
var three = require('three');
var cannon = require('cannon');

var duration = 500;

var geometry = new three.SphereGeometry(1, 16, 8);
function EnergyBubblePopMesh() {
	var material = new three.MeshBasicMaterial({
		transparent: true,
		opacity: 0.5,
		depthWrite: false
	});
	three.Mesh.call(this, geometry, material);
}

EnergyBubblePopMesh.prototype = Object.create(three.Mesh.prototype);

function EnergyBubblePop(pos, size, onComplete) {
	var mesh = new EnergyBubblePopMesh();
	mesh.position.copy(pos);
	mesh.scale.multiplyScalar(size);

	this.mesh = mesh;
	this.material = mesh.material;
	this.size = size;
	this.timeStart = performance.now();

	this.onComplete = onComplete;
	this.onEnterFrame = onEnterFrame.bind(this);
	this.onEnterFrame();
}

function onEnterFrame() {
	var timePassed = performance.now() - this.timeStart;
	if(timePassed > duration) {
		this.onComplete(this);
		return;
	} else {
		var ratio = timePassed / duration;
		var invRatio = 1 - ratio;
		this.material.opacity = invRatio;
		this.material.color.setRGB(1, invRatio*2 - 0.5, invRatio*invRatio);
		var s = this.size * (0.1 + ratio);
		this.mesh.scale.set(s, s, s);
	}
}

module.exports = EnergyBubblePop;
