// gameObjects/tools/BallMaker.js
var three = require('three');
var CANNON = require('cannon');
var Signal = require('signals').Signal;
var CheckerboardTexture = require('threejs-texture-checkerboard');
var CollisionLayers = require('CollisionLayers');

function BallMaker(world, pos) {
	
	// bodyShape.material = groundMaterial;
	var toolBody = new CANNON.Body({
		mass: 50, // kg 
		fixedRotation: true,
		linearDamping: 0.5,
		resistGravity: true,
		collisionFilterGroup: CollisionLayers.ITEMS,
		collisionFilterMask: CollisionLayers.ENVIRONMENT | CollisionLayers.PLAYER | CollisionLayers.ITEMS
	});
	// toolBody.resistGravity = true;
	var bodySphereRecipe = [
		[0, 1, 0, 0.5]
	];
	for(var i = 0; i < bodySphereRecipe.length; i++) {
		var d = bodySphereRecipe[i];
		var shape = new CANNON.Sphere(d[3]);
		toolBody.addShape(shape, new CANNON.Vec3(d[0], d[1], d[2]));
	}
	toolBody.quaternion.setFromEuler(0, 0, 0);
	toolBody.position.set(pos.x, pos.y, pos.z);
	var toolMaterial = new THREE.MeshPhongMaterial({
		map: new CheckerboardTexture(0x000000, 0xffffff)
	});
	var toolMesh = new THREE.Mesh(
		new THREE.CylinderGeometry(0.125, 0.125, 1, 16, 1),
		toolMaterial
	);

	toolBody.interactiveObject = { type: "tool", object: this };

	var pivot = new THREE.Object3D();
	pivot.add(toolMesh);

	this.world = world;
	this.mesh = pivot;
	this.subMesh = toolMesh;
	this.body = toolBody;
	this.bodySphereRecipe = bodySphereRecipe;

	this.onEnterFrame = onEnterFrame.bind(this);
	this.onUpdateSim = onUpdateSim.bind(this);
	this.primaryFire = primaryFire.bind(this);
}

var spinSpeed = 0.05;
function onEnterFrame() {
	this.subMesh.rotation.z += spinSpeed;
}

function onUpdateSim() {
	var toolBody = this.body;
}

function primaryFire(pos, playerSize) {
	this.world.makeBall(pos.x, pos.y, pos.z, playerSize);
}

module.exports = BallMaker;
